flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/api/ZkLockFactory.java

```diff
@@ -0,0 +1,170 @@
+   * @param sessionTimeoutMs Session timeout in milliseconds
+   * @param connectionTimeoutMs Connection timeout in milliseconds
+   * @param baseSleepTimeMs Base sleep time in milliseconds
+   * @param maxRetries Maximum number of retries
+   */
+  public ZkLockFactory(
+      String connectString,
+      String lockId,
+      int sessionTimeoutMs,
+      int connectionTimeoutMs,
+      int baseSleepTimeMs,
+      int maxRetries) {
+    Preconditions.checkNotNull(connectString, "Zookeeper connection string cannot be null");
+    Preconditions.checkNotNull(lockId, "Lock ID cannot be null");
+    this.connectString = connectString;
+    this.lockId = lockId;
+    this.sessionTimeoutMs = sessionTimeoutMs;
+    this.connectionTimeoutMs = connectionTimeoutMs;
+    this.baseSleepTimeMs = baseSleepTimeMs;
+    this.maxRetries = maxRetries;
+  }
+
+  @Override
+  public void open() {
+    this.client =
+        CuratorFrameworkFactory.builder()
+            .connectString(connectString)
+            .sessionTimeoutMs(sessionTimeoutMs)
+            .connectionTimeoutMs(connectionTimeoutMs)
+            .retryPolicy(new ExponentialBackoffRetry(baseSleepTimeMs, maxRetries))
+            .build();
+    client.start();
+
+    try {
+      if (!client.blockUntilConnected(connectionTimeoutMs, TimeUnit.MILLISECONDS)) {
+        throw new IllegalStateException("Connection to Zookeeper timed out");
+      }
+
+      this.taskSharedCount = new SharedCount(client, LOCK_BASE_PATH + lockId + "/task", 0);
+      this.recoverySharedCount = new SharedCount(client, LOCK_BASE_PATH + lockId + "/recovery", 0);
+      taskSharedCount.start();
+      recoverySharedCount.start();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted while connecting to Zookeeper", e);
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to initialize SharedCount", e);
+    }
+  }
+
+  @Override
+  public Lock createLock() {
+    return new ZkLock(taskSharedCount);
+  }
+
+  @Override
+  public Lock createRecoveryLock() {
+    return new ZkLock(recoverySharedCount);
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      if (taskSharedCount != null) {
+        taskSharedCount.close();
+      }
+
+      if (recoverySharedCount != null) {
+        recoverySharedCount.close();
+      }
+    } finally {
+      if (client != null) {
+        client.close();
+      }
+    }
+  }
+
+  /** Zookeeper lock implementation */
+  private static class ZkLock implements Lock {
+    private final SharedCount sharedCount;
+
+    private ZkLock(SharedCount sharedCount) {
+      this.sharedCount = sharedCount;
+    }
+
+    @Override
+    public boolean tryLock() {
+      if (isHeld()) {
+        LOG.debug("Lock is already held for {}", this);
+        return false;
+      }
+
+      try {
+        return sharedCount.trySetCount(sharedCount.getVersionedValue(), 1);
```

Comment:
I'm not comfortable with this one.

Shouldn't we get/check/trySet? This way, if there is a concurrent lock between `isHeld` and the `trySet`, then we just ignore it.